pkg/gds/admin_test.go

```diff
@@ -0,0 +1,966 @@
+		StandardClaims: jwt.StandardClaims{
+			Id:        uuid.NewString(),
+			Audience:  "http://localhost",
+			IssuedAt:  time.Now().Unix(),
+			NotBefore: time.Now().Unix(),
+			ExpiresAt: time.Now().Unix(),
+		},
+	}
+	accessToken := jwt.NewWithClaims(jwt.SigningMethodRS256, claims)
+	refreshToken, err := s.tokens.CreateRefreshToken(accessToken)
+	require.NoError(err)
+	access, err := s.tokens.Sign(accessToken)
+	require.NoError(err)
+	refresh, err := s.tokens.Sign(refreshToken)
+	require.NoError(err)
+
+	// Missing access token
+	request := &httpRequest{
+		method: http.MethodPost,
+		path:   "/v2/reauthenticate",
+		in: &admin.AuthRequest{
+			Credential: refresh,
+		},
+	}
+	c, w := s.makeRequest(request)
+	res := s.doRequest(a.Reauthenticate, c, w, nil)
+	require.Equal(http.StatusUnauthorized, res.StatusCode)
+
+	// Invalid access token
+	request.headers = map[string]string{
+		"Authorization": "Bearer invalid",
+	}
+	c, w = s.makeRequest(request)
+	res = s.doRequest(a.Reauthenticate, c, w, nil)
+	require.Equal(http.StatusUnauthorized, res.StatusCode)
+
+	// Missing refresh token
+	request.in = &admin.AuthRequest{}
+	request.headers = map[string]string{
+		"Authorization": "Bearer " + access,
+	}
+	c, w = s.makeRequest(request)
+	res = s.doRequest(a.Reauthenticate, c, w, nil)
+	require.Equal(http.StatusUnauthorized, res.StatusCode)
+
+	// Invalid refresh token
+	request.in = &admin.AuthRequest{
+		Credential: "invalid",
+	}
+	c, w = s.makeRequest(request)
+	res = s.doRequest(a.Reauthenticate, c, w, nil)
+	require.Equal(http.StatusUnauthorized, res.StatusCode)
+
+	// Mismatched access and refresh tokens
+	claims.Id = uuid.NewString()
+	otherToken := jwt.NewWithClaims(jwt.SigningMethodRS256, claims)
+	other, err := s.tokens.Sign(otherToken)
+	require.NoError(err)
+	request.in = &admin.AuthRequest{
+		Credential: refresh,
+	}
+	request.headers = map[string]string{
+		"Authorization": "Bearer " + other,
+	}
+	c, w = s.makeRequest(request)
+	res = s.doRequest(a.Reauthenticate, c, w, nil)
+	require.Equal(http.StatusUnauthorized, res.StatusCode)
+
+	// Successful reauthentication
+	request.in = &admin.AuthRequest{
+		Credential: refresh,
+	}
+	request.headers = map[string]string{
+		"Authorization": "Bearer " + access,
+	}
+	c, w = s.makeRequest(request)
+	res = s.doRequest(a.Reauthenticate, c, w, nil)
+	require.Equal(http.StatusOK, res.StatusCode)
+	// Double cookie tokens should be set
+	cookies := res.Cookies()
+	require.Len(cookies, 2)
+	for _, cookie := range cookies {
+		require.Equal(s.conf.Admin.CookieDomain, cookie.Domain)
+	}
+}
+
+// Test that the Summary endpoint returns the correct response.
+func (s *gdsTestSuite) TestSummary() {
+	require := s.Require()
+	a := s.initAdmin(s.dbPath)
+
+	request := &httpRequest{
+		method: http.MethodGet,
+		path:   "/v2/summary",
+	}
+	actual := &admin.SummaryReply{}
+	c, w := s.makeRequest(request)
+	rep := s.doRequest(a.Summary, c, w, actual)
+	require.Equal(http.StatusOK, rep.StatusCode)
```

Comment:
Ah, I see - so we're circumventing the middleware by going directly to the handler. Authenticate and CSRF protect are never called. That makes a lot of sense. 

Could we create stories to test endpoints for authenticate and csrf protection required?

Reply:
Sure, my goal for these specific tests was to make sure the logic within the handlers is correct when given a context, but I agree we do need to make sure that the authenticated and protected endpoints work from a more external perspective.